File: solutions/csharp/need-for-speed/1/NeedForSpeed.cs
class RemoteControlCar
{    
    private int _speed;
    private int _batteryDrain;
    
    private int _distanceDriven = 0;
    private int _batteryPercentage = 100;
    
    public RemoteControlCar(int speed, int batteryDrain){
        this._speed = speed;
        this._batteryDrain = batteryDrain;
    }

    public bool BatteryDrained()
    {
        if ((this._batteryPercentage - this._batteryDrain < 0) || this._batteryPercentage == 0) return true;
        else return false;
    }

    public int DistanceDriven()
    {
        return this._distanceDriven;
    }

    public void Drive()
    {
        if (BatteryDrained()) return;
        this._distanceDriven += this._speed;
        this._batteryPercentage -= this._batteryDrain;
    }

    public static RemoteControlCar Nitro()
    {
        return new RemoteControlCar(50, 4);
    }
}

class RaceTrack
{
    private int _distance;

    public RaceTrack (int distance) {
        this._distance = distance;
    }

    public bool TryFinishTrack(RemoteControlCar car)
    {
        while (!car.BatteryDrained()) {
            car.Drive();
        }
        return car.DistanceDriven() >= this._distance;
    }
}

File: solutions/java/need-for-speed/1/src/main/java/NeedForSpeed.java
class NeedForSpeed {

    private int speed;
    private int batteryDrain;
    private int battery = 100;
    private int distance = 0;
    
    NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public boolean batteryDrained() {
        return (this.battery - batteryDrain < 0) || this.battery == 0;
    }

    public int distanceDriven() {
       return this.distance;
    }

    public void drive() {
        if (batteryDrained()) return;
        this.distance += this.speed;
        this.battery -= batteryDrain;
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {
    private int distance; 
    RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean canFinishRace(NeedForSpeed car) {
        while (car.distanceDriven() < this.distance && !car.batteryDrained())
        {
            car.drive();
        }
        return car.distanceDriven() >= this.distance;
    }
}
